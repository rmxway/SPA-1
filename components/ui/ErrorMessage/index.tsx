import { FC } from 'react';

import { WrapperError } from './styled';

interface ErrorMessageProps {
	error: string;
}

export const ErrorMessage: FC<ErrorMessageProps> = ({ error }) => (
	<WrapperError>{error && <span>{error}</span>}</WrapperError>
);
export default ErrorMessage;
