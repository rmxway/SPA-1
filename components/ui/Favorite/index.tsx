import { Wrapper } from './styled';

interface FavoriteProps {
	active?: boolean;
	onActive?: () => void;
}

export function Favorite({ active = false, onActive, ...props }: FavoriteProps) {
	return (
		<Wrapper {...props} {...{ active }} onClick={onActive}>
			{!active ? <i className="icofont icofont-favorite-fill" /> : <i className="icofont icofont-favorite" />}
		</Wrapper>
	);
}

export default Favorite;
