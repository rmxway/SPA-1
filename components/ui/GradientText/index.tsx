import { FC } from 'react';

import { Gradients } from '@/@types/styled';

import { GradientTextStyled } from './styled';

interface GradientTextProps extends React.HTMLAttributes<HTMLDivElement> {
	size?: number;
	gradient?: keyof Gradients;
}

/**
 *
 * @param gradient(Gradients) - main, dark
 * @param size(number) - size in pixels
 * @returns
 */

export const GradientText: FC<GradientTextProps> = ({ size, gradient = 'main', children, ...props }) => (
	<GradientTextStyled $size={size} $gradient={gradient} {...props}>
		{children}
	</GradientTextStyled>
);
export default GradientText;
