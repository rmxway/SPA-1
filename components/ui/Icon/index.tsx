import { ElementType } from 'react';

import { Icofont } from '@/services';

interface IconProps {
	as?: ElementType;
	icon: Icofont;
	size?: number;
	className?: string;
}

export const Icon = ({ as: Tag = 'i', icon = 'cart', size, className, ...props }: IconProps) => (
	<Tag className={`icofont icofont-${icon} ${className}`} {...props} style={{ fontSize: size }} />
);
