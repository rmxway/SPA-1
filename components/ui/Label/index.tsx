import { FC } from 'react';

import { LabelStyled } from './styled';

interface LabelProps {
	label: string;
	name: string;
}

export const Label: FC<LabelProps> = ({ label, name }) => <LabelStyled htmlFor={name}>{label}</LabelStyled>;

export default Label;
