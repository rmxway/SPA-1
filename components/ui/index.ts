export * from './Button';
export * from './Favorite';
export * from './Icon';
export * from './Input';
export * from './LinkIcon';
export * from './Loader';
export * from './Modal';
export * from './Sticker';
export * from './TopBlock';
