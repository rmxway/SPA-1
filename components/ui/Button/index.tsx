import { AnimatePresence, motion } from 'framer-motion';
import { FC, PropsWithChildren } from 'react';

import { Icon } from '@/components/ui/Icon';
import icons from '@/public/assets/fonts/icofont/icofont.json';

import { ButtonStyle, Props } from './styled';

const textVar = {
	hidden: { opacity: 0 },
	visible: { opacity: 1 },
};

type ButtonProps = Props & PropsWithChildren & { animate?: boolean; icon?: keyof typeof icons };

/**
 * Custom button with some properties.
 ** Необходимо доработать анимацию
 * @param {?boolean} $w100 - width 100%
 * @param {?boolean} $inactive - pointer-events: none
 * @param {?boolean} $margins - margin-right and margin-bottom 10px
 * @param {?boolean} animate - animate inner text (not working)
 * @param {?boolean} icon - select icon
 * @param {?boolean} (styled boolean params) - $success, $danger, $primary, $white, $dark
 */

export const Button: FC<ButtonProps> = ({ children, animate = false, icon, ...props }) => {
	const memoText = typeof children === 'string' ? children.split('') : [''];

	return (
		<ButtonStyle {...props}>
			{animate ? (
				<AnimatePresence mode="wait">
					<motion.div layout variants={textVar} initial="hidden" animate="visible" exit="hidden">
						{memoText?.map((item, idx) => (
							<motion.span
								layout
								variants={textVar}
								transition={{ delay: 0.1 * idx }}
								key={Number(Math.random())}
							>
								{item}
							</motion.span>
						))}
					</motion.div>
				</AnimatePresence>
			) : (
				children
			)}
			{icon && <Icon {...{ icon }} />}
		</ButtonStyle>
	);
};

export default Button;
