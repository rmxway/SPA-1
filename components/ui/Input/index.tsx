import React, { forwardRef } from 'react';
import { IMaskInput, ReactElementProps } from 'react-imask';

import { ErrorMessage } from '@/components/ui/ErrorMessage';
import { Label } from '@/components/ui/Label';

import { InputWrapper } from './styled';

export interface InputProps {
	label?: string;
	mask?: string;
	name: string;
	error?: string;
	success?: boolean;
	noPadding?: boolean;
	className?: string;
}

type ReactInputProps = ReactElementProps<HTMLInputElement>;

export const Input = forwardRef<HTMLInputElement, InputProps & React.InputHTMLAttributes<HTMLInputElement>>(
	({ name, label, mask, error, success, noPadding, className, ...props }, ref) => (
		<InputWrapper {...{ className }} $error={!!error} $success={success} $noPaddings={noPadding}>
			{label && <Label {...{ label, name }} />}
			{mask ? (
				<IMaskInput
					{...(props as ReactInputProps)}
					{...{ name, mask }}
					type="text"
					autoComplete="off"
					inputRef={ref}
				/>
			) : (
				<input {...props} {...{ ref, name }} id={name} type="text" autoComplete="off" />
			)}

			{error && <ErrorMessage {...{ error }} />}
		</InputWrapper>
	),
);

export default Input;
