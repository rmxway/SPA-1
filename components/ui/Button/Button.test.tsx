import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from '@/components/ui';
import { StyledComponentsRegistry } from '@/lib/registry';

describe('Button/UI:', () => {
	it('Render is fine', async () => {
		const user = userEvent.setup();
		const clickFn = jest.fn();

		const { getByRole, container } = render(
			<StyledComponentsRegistry isJest>
				<Button onClick={clickFn}>Button</Button>
			</StyledComponentsRegistry>,
		);

		const button = getByRole('button');

		expect(button).toBeInTheDocument();
		expect(container).toMatchSnapshot();

		await user.click(button);

		expect(clickFn).toHaveBeenCalledTimes(1);
	});

	it('$inactive is fine', () => {
		const { getByRole } = render(
			<StyledComponentsRegistry isJest>
				<Button $inactive>Inactive Button</Button>
			</StyledComponentsRegistry>,
		);

		expect(getByRole('button')).toHaveStyle('pointer-events: none');
	});

	it('Icon is valid', () => {
		const icon = 'cart';
		const { container } = render(
			<StyledComponentsRegistry isJest>
				<Button icon={icon}>Cart</Button>
			</StyledComponentsRegistry>,
		);

		const iconElement = container.getElementsByTagName('i')[0];
		expect(iconElement.className.includes(`icofont-${icon}`)).toBeTruthy();
	});
});
