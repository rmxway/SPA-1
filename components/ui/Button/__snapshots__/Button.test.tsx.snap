// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button/UI: Render is fine 1`] = `
<div>
  <button
    class="sc-6ffd862-0 ljzhXT"
  >
    Button
  </button>
</div>
`;
